memcpy-based fast, typed shuffle.

anntzer opened the pull request:
Only for ndarrays exactly, as subtypes (e.g. masked arrays) may not
like directly shuffling the underlying buffer (in fact, the old
implementation destroyed the underlying values of masked arrays while
shuffling).

Also handles struct-containing-object 1d ndarrays properly.

See #6776 for an earlier, less general (but even faster: ~6x)
improvement attempt, #5514 for the original issue.

anntzer committed:
MAINT: memcpy-based ~4x faster, typed shuffle.

Only for 1d-ndarrays exactly, as subtypes (e.g. masked arrays) may not
allow direct shuffle of the underlying buffer (in fact, the old
implementation destroyed the underlying values of masked arrays while
shuffling).

Also handles struct-containing-object 1d ndarrays properly.

See #6776 for an earlier, less general (but even faster: ~6x)
improvement attempt, #5514 for the original issue.

charris commented:
Once you have this working, could you add a note to `doc/release/1.11.0-notes.rst`? The commit message should also begin `MAINT: ` as you are improving existing functionality.

anntzer commented:
Fixed.
Note that while this approach is more general than the one in #6776 and provides 4x speedup, the specializations in #6776 were even faster (~6x speedup).  Perhaps it is possible to coerce gcc into generating specializations of `shuffle` for common `nbytes` sizes (1/2/4/8)?

homu commented:
:umbrella: The latest upstream changes (presumably #6932) made this pull request unmergeable. Please resolve the merge conflicts.

anntzer committed:
Top shuffle speed for machine-sized ints/floats.

Apparently gcc only specializes one branch (the last one) so I went for
another 33% performance increase (matching #6776) in what's likely the
most common use case.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4967,33 +4969,53 @@ cdef class RandomState:
+            # Fast, statically typed path: shuffle the underlying buffer.
+            # Only for non-empty, 1d objects of class ndarray (subclasses such
+            # as MaskedArrays may not support this approach).
+            x_ptr = <char*><size_t>x.ctypes.data
```
Why the cast to `size_t`?

anntzer replied:
Because Cython interprets `<char*>python_object` as casting a bytes object (*not* an object supporting the buffer protocol) to a `char*`, but `<char*>c_object` as a C-style cast.  (If you try it you'll get "TypeError: expected bytes, int (or whatever you pass in) found")

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4967,33 +4969,53 @@ cdef class RandomState:
                     x[i], x[j] = x[j], x[i]
-                    i = i - 1
+
+    cdef inline _shuffle_raw(self, npy_intp n, npy_intp nbytes, npy_intp stride,
```
Does `inline` buy you anything here?

anntzer replied:
Yes, without it GCC doesn't inline this branch and it is actually slower that the nonspecialized branch.

charris commented on numpy/random/tests/test_random.py:
```diff
@@ -205,12 +208,13 @@ def test_repeatability(self):
 
 
-class TestRandomDist(TestCase):
+class TestRandomDist:
```
This will be an old style class on Python 2.7, best derive from object if not `TestCase`. Any reason to avoid `TestCase`?

anntzer replied:
Because subclasses of TestCase do not support generative tests (http://nose.readthedocs.org/en/latest/writing_tests.html#test-generators)

charris commented on numpy/random/tests/test_random.py:
```diff
@@ -205,12 +208,13 @@ def test_repeatability(self):
 
-    def setUp(self):
-        self.seed = 1234567890
+    @classmethod
```
Why `classmethod`?

anntzer replied:
Because a regular `setUp` method would not be invoked specifically by the generative test.

charris replied:
What does making this a generative test buy you here? Why not just call the assertion in the loop as before?

anntzer replied:
With a generative test, in case of a test failure, you know immediately which iteration failed; additionally even if that happens the following iterations of the loop are still executed (and additional failures can be reported too).

matthew-brett replied:
At least in our code, we used generator tests heavily, then ended up swapping them back to normal tests, because they are considerably more difficult to debug.

In practice, I don't think it buys you much if one test fails but the rest in the same function still run - anyone debugging will end up fixing the first failure and then re-running the tests in any case.

njsmith replied:
I share @matthew-brett and @charris's vague dislike of generator-based tests. It's a nice idea and does work, but just a bit too clunky in practice.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4967,33 +4969,53 @@ cdef class RandomState:
-            # rows; we need a bounce buffer.
+        cdef:
+            npy_intp i, j, n = len(x)
+            size_t stride, nbytes
```
We generally use `npy_intp`, not `size_t`. Note that arrays will be in a valid range for `npy_intp`.

anntzer replied:
Fixed.

charris commented on numpy/random/tests/test_random.py:
```diff
@@ -352,40 +356,41 @@ def test_bytes(self):
         np.testing.assert_equal(actual, desired)
 
     def test_shuffle(self):
```
ISTR that we don't like generators like this for testing. @rgommers Do you remember what that was about?

argriffing replied:
From a comment by @pv https://github.com/scipy/scipy/pull/3719#discussion_r13781425 one reason is that it's harder to debug failing tests, for example

```python
import numpy as np
from numpy.testing import assert_allclose

def test_foo():
    A = np.arange(100)
    B = np.concatenate([A, A])
    yield assert_allclose, A, B
```

gives a failing test result that looks like this:

```
$ nosetests                           
F
======================================================================
FAIL: test.test_foo(array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
----------------------------------------------------------------------
Traceback (most recent call last):
[snip]
```

which looks messy.

anntzer replied:
You can set the `description` attribute of the yielded test to improve the output, e.g.
```
import numpy as np
from numpy.testing import assert_allclose

def test_foo():

    def allclose(description):
        f = lambda x, y: assert_allclose(x, y)
        f.description = description
        return f

    for i in range(100):
        A = np.arange(100)
        if i != 50:
            B = A
        else:
            B = np.arange(200)
        yield allclose("test_foo:test #{}".format(i)), A, B

def test_bar():
    for i in range(100):
        A = np.arange(100)
        if i != 50:
            B = A
        else:
            B = np.arange(200)
        assert_allclose(A, B)
```
gives, IMO, a nice output:
```
..................................................F.................................................F
======================================================================
FAIL: test_foo:test #50
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/tmp/test_foo.py", line 7, in <lambda>
    f = lambda x, y: assert_allclose(x, y)
  File "/usr/lib/python3.5/site-packages/numpy/testing/utils.py", line 1347, in assert_allclose
    verbose=verbose, header=header)
  File "/usr/lib/python3.5/site-packages/numpy/testing/utils.py", line 663, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

(shapes (100,), (200,) mismatch)
 x: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,...
 y: array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,...

======================================================================
FAIL: test_foo.test_bar
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/tmp/test_foo.py", line 26, in test_bar
    assert_allclose(A, B)
  File "/usr/lib/python3.5/site-packages/numpy/testing/utils.py", line 1347, in assert_allclose
    verbose=verbose, header=header)
  File "/usr/lib/python3.5/site-packages/numpy/testing/utils.py", line 663, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

(shapes (100,), (200,) mismatch)
 x: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,...
 y: array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,...

----------------------------------------------------------------------
Ran 101 tests in 0.056s

FAILED (failures=2)
```

matthew-brett replied:
The description is not the issue with debugging, it's the context when you enable ``--pdb-failure`` or ``--ipdb-failure`` - this is what many of us use for debugging failed tests.

anntzer replied:
I don't see any context using either `--pdb-failure` or `--ipdb-failure` (not that I knew about these options, thanks for pointing them out).

matthew-brett replied:
I'm not sure what you mean, but I think you'll find that non-generator tests will drop you into the errored test, and you can `u` you way to the enclosing test function to inspect the stack, but you can't do that with the `yield` test functions.

anntzer replied:
Oh, I see.  I'll switch this out then.
I guess nose should try to stitch in a traceback to the still running generator (`gen.gi_frame` anyone?).  Let's open a bug there...

anntzer commented:
(switched back to non-generative test)

anntzer committed:
Revert to non-generative test.

homu commented:
:umbrella: The latest upstream changes (presumably #6453) made this pull request unmergeable. Please resolve the merge conflicts.

anntzer commented:
Kindly bumping this issue and hoping it'll make it for 1.11.0.

njsmith commented:
You have conflicts?

anntzer commented:
It's the 3rd time (including the previous PR) I'm merging trivial conflicts in the release notes so I'm only going to do it if someone tells me this will be merged right after.

njsmith commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4967,33 +4969,52 @@ cdef class RandomState:
+            # as MaskedArrays may not support this approach).
+            x_ptr = <char*><size_t>x.ctypes.data
+            stride = x.strides[0]
+            nbytes = x[:1].nbytes
```
This confused me for a while. I think the more conventional way to write this would be `x.dtype.itemsize`.

anntzer replied:
I think I wrote this back when I hoped to be able to handle nd-arrays there too (this can't work because the sub-arrays are not necessarily contiguous).  Changing it back.

njsmith commented:
One nitpick, but otherwise LGTM. @charris, any objections to merging for 1.11?

charris commented:
@njsmith I haven't reviewed it, but I'm happy for any help in that regard. We have 123 PRs pending, so if it looks good to you, put it in. It might be another day or two before the 1.11 branch.

njsmith commented:
Let's try it out then. Thanks Antony!
